### PR TITLE
Adds the app tree to the controller overview

### LINKED ISSUE #163: Summary application tree
Requires #160.

A simple visual summary of the hierarchy of running controllers and DAQ processes. The data for this should be accessible via the [ls command](https://github.com/DUNE-DAQ/drunc/wiki/Controller#ls) of the root controller. A simple way of rendering this for the time being would be as nested lists.

- [x] #228
- [x] #229

### LINKED ISSUE #228: Implement front end side of things in the application tree summary
Using web components and a placeholder application tree for now. See https://github.com/ImperialCollegeLondon/drunc_ui/issues/163#issuecomment-2431363673 and links therein.

### PR DESCRIPTION
**Pull request by dalonsoa**
# Description

Adds the application tree to the controller overview page, as well as the required functions to retrieve the information from the backend. 

NOTE: Ignore the lock. The issue it was meant to fixed is being fixed properly in #263 and will be incorporated into this one when merged.

https://github.com/user-attachments/assets/c7803cee-17b2-4415-bedb-d3cb6688fd16

Fixes #228 
Fixes #163

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (eg. `python -m pytest`)
- [x] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [x] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

**Comment by AdrianDAlessandro** [comment]
Can you change the target branch of this PR to `fsm_dialog`?

**Comment by dalonsoa** [comment]
Done! Sorry for the mistake

**Comment by dalonsoa** [comment]
I also forgot to upload the screen recording 🤦‍♂️ Will do first thing tomorrow morning.

File: controller/templates/controller/partials/app_tree.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<sl-tree class="custom-icons">
+<sl-icon name="plus-square" slot="expand-icon"></sl-icon>
+<sl-icon name="dash-square" slot="collapse-icon"></sl-icon>
+{{ tree }}
+</sl-tree>
+<style>
+        .custom-icons sl-tree-item::part(expand-button) {
+          /* Disable the expand/collapse animation */
```
Comment (cc-a) [inline controller/templates/controller/partials/app_tree.html:8]: What do you have against the expand/collapse animation?

**Review by cc-a — CHANGES_REQUESTED** [review]
Looking good. Just a few refinements needed.

File: controller/controller_interface.py [hunk 2]
```diff
@@ -21,6 +22,9 @@
 }
 """Mapping of argument types to their protobuf message types."""
 
+connectivity_lock = Lock()
```
[the remark this answers is not in the document]
Comment (dalonsoa) [reply controller/controller_interface.py:25]: Shall we do this a separate PR?

File: controller/templates/controller/partials/app_tree.html (hunk 1 above)
Comment (dalonsoa) [reply controller/templates/controller/partials/app_tree.html:8]: I don't like them, so I guess just an opinionated choice. Can change it if there's appetite for animations.

File: controller/controller_interface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalonsoa) [reply controller/controller_interface.py:25]: I'm opening a separate PR with this bit of code - after checking that also works for me!

File: controller/views/partials.py [hunk 3]
```diff
@@ -53,3 +54,31 @@ def dialog(request: HttpRequest) -> HttpResponse:
...
+    Returns:
+        The shoelace tree component as safe HTML code.
+    """
+    return mark_safe(
+        f"<sl-tree-item expanded> {app['name']}"
+        + "".join(
+            app_to_shoelace_tree(cast(ci.AppType, child)) for child in app["children"]
+        )
+        + "</sl-tree-item>"
+    )
```
[the remark this answers is not in the document]
Comment (dalonsoa) [reply controller/views/partials.py]: Sorted

**Review by cc-a — APPROVED** [review]
Great stuff. One more on the stack.

**Review by AdrianDAlessandro — APPROVED** [review]
LGTM once the threading lock is removed

File: controller/templates/controller/index.html [hunk 4]
```diff
@@ -25,6 +25,12 @@
     }
   </style>
 {% endblock extra_css %}
+{% block extra_js %}
+  <link rel="stylesheet"
+        href="https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.18.0/cdn/themes/light.css" />
+  <script type="module"
+          src="https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.18.0/cdn/shoelace-autoloader.js"></script>
+{% endblock extra_js %}
```
Comment (jamesturner246) [inline controller/templates/controller/index.html:33]: We will need local copies of the js and css files for deployment.

**Review by jamesturner246 — CHANGES_REQUESTED** [review]
Looks good, besides needing local copies of the js and css. 

I understand a lot of this will change in #262.

File: controller/templates/controller/index.html (hunk 4 above)
Comment (dalonsoa) [reply controller/templates/controller/index.html:33]: I've mentioned it before and I say it here again - manually coping JS files to the repo I think is a bad idea. It's a recipe for getting those components out of sync and outdated. We should use a proper dependency management for JS packages, like `npm`, now that it seems we have a few.

Comment (dalonsoa) [reply controller/templates/controller/index.html:33]: Besides, I don't know how to download this locally and ship it with the software. Shoelace seems to be a complex tool with multiple components and style files, not a simple, self-contained JS file.

I'd really suggest to leave this as it is an open an issue to move to using `npm` (or other tool) for JS dependency management, now that it seems we are going to have a gap until there're more things to do in the project.

Comment (jamesturner246) [reply controller/templates/controller/index.html:33]: Sure, no problem. Just making aware that they will not be able to be remotely sourced in deployment. Let's leave it then.

**Review by jamesturner246 — APPROVED** [review]

File: controller/templates/controller/index.html (hunk 4 above)
Comment (cc-a) [reply controller/templates/controller/index.html:33]: Add #275 for this

**Pull request merged**